### DISCUSSION TITLE
LocalStack 3.2.0 with AWS Spring Cloud 2.4.x?

### DISCUSSION THREAD

**Discussion by pcalouche** [awspring/spring-cloud-aws#1120] [Q&A] [answered]
Question

I realize version `2.4.x` is no longer supported, but I was curious if it can work with LocalStack 3 since we're transitioning between AWS Spring Cloud versions at my company. I tried the following from the `2.4.x` branch:

Verified the [SqsSampleApplicationTests](https://github.com/awspring/spring-cloud-aws/blob/2.4.x/spring-cloud-aws-samples/spring-cloud-aws-sqs-sample/src/test/java/io/awspring/cloud/sqs/sample/SqsSampleApplicationTests.java) successfully runs before making any changes.

Then I set AWS Java SDK version to `1.12.583` in https://github.com/awspring/spring-cloud-aws/blob/2.4.x/spring-cloud-aws-dependencies/pom.xml#L35

```xml
<aws-java-sdk.version>1.12.583</aws-java-sdk.version>
```

Updated LocalStack image in [SqsSampleApplicationTests](https://github.com/awspring/spring-cloud-aws/blob/2.4.x/spring-cloud-aws-samples/spring-cloud-aws-sqs-sample/src/test/java/io/awspring/cloud/sqs/sample/SqsSampleApplicationTests.java#L48) to:

```java
@Container
static LocalStackContainer localstack = new LocalStackContainer(DockerImageName.parse("localstack/localstack:3.2.0")).withServices(SQS).withReuse(true);
```

When running the test I see the following error:

`Caused by: org.apache.http.conn.HttpHostConnectException: Connect to sqs.us-east-1.localhost.localstack.cloud:4566 [sqs.us-east-1.localhost.localstack.cloud/127.0.0.1] failed: Connection refused (Connection refused)`

I experienced the same behavior in my applications tests when switching to LocalStack `3.2.0`.

That URL looks odd to me. I appreciate any insight you can provide.

**Comment by maciejwalkowiak** [comment]
I think it is more a question about which version of AWS SDK works with which LocalStack version. Perhaps folks from LocalStack can help you?

I'll convert it into discussion since it is not really an issue in Spring Cloud AWS.

**Reply by pcalouche** [reply, marked as answer]
@maciejwalkowiak Your suggestion to ask the LocalStack folks was a good one. It turned out there is an active issue about this and it wasn't a case of an AWS Java SDK mismatch but rather a LocalStackContainer 3.x.x config change.

Setting the `SQS_ENDPOINT_STRATEGY` environment variables to `dynamic` resolves the URLs correctly in the test example I mentioned above, and in all my application's tests that were failing. The issue reference is https://github.com/localstack/localstack/issues/10024.

```java
@Container
	static LocalStackContainer localstack = new LocalStackContainer(
			DockerImageName.parse("localstack/localstack:3.3.0"))
		.withServices(SQS)
		.withReuse(true)
		.withEnv("SQS_ENDPOINT_STRATEGY", "dynamic");
```
